### PR TITLE
msg/async/dpdk:add the handling of DPDK initialization failure

### PR DESCRIPTION
**Pull request by fengchunsong**
If rte_eal_init returns with failure,the waiting msgr-worker thread is
woken up for exception handling.

Signed-off-by: Chunsong Feng <fengchunsong@huawei.com>
Reviewed-by: luo rixin <luorixin@huawei.com>
Reviewed-by: Han Fengzhe  <hanfengzhe@hisilicon.com>

### CODE REVIEW

**Comment by fengchunsong** [comment]
@tchaikov please help to review,thanks

**Comment by fengchunsong** [comment]
jenkins test make check

**Comment by fengchunsong** [comment]
@cyx1231st please help to review,thanks.

**Comment by fengchunsong** [comment]
@tchaikov please help to review,thanks.

File: src/msg/async/dpdk/dpdk_rte.cc [hunk 1]
```diff
@@ -137,8 +137,12 @@ namespace dpdk {
       if (!rte_initialized) {
         /* initialise the EAL for all */
         int ret = rte_eal_init(cargs.size(), cargs.data());
-        if (ret < 0)
-          return;
+	if (ret < 0) {
+          std::unique_lock<std::mutex> l(lock);
```
Comment (tchaikov) [inline src/msg/async/dpdk/dpdk_rte.cc:141]:
```suggestion
          std::unique_lock locker(lock);
```

please avoid using a single `l` for the variable name. `l` looks like `1` with a bad font.

File: src/msg/async/dpdk/dpdk_rte.cc [hunk 2]
```diff
@@ -159,7 +163,7 @@ namespace dpdk {
     std::unique_lock<std::mutex> l(lock);
     while (!done)
       cond.wait(l);
-    return 0;
+    return initialized ? 0:-EIO;
```
Comment (tchaikov) [inline src/msg/async/dpdk/dpdk_rte.cc:166]: nit, not quite related to this change, but the code could be refactored so it reads:
```c++
std::unique_lock locker(lock);
cond.wait(locker, [&] { return done; });
return initialized ? 0 : -EIO;
```

Comment (tchaikov) [reply src/msg/async/dpdk/dpdk_rte.cc:166]: please add spaces around `:`.

File: src/msg/async/dpdk/dpdk_rte.cc (hunk 1 above)
Comment (fengchunsong) [reply src/msg/async/dpdk/dpdk_rte.cc:141]: Ok

File: src/msg/async/dpdk/dpdk_rte.cc (hunk 2 above)
Comment (fengchunsong) [reply src/msg/async/dpdk/dpdk_rte.cc:166]: Ok

Comment (fengchunsong) [reply src/msg/async/dpdk/dpdk_rte.cc:166]: > nit, not quite related to this change, but the code could be refactored so it reads:
> 
> ```c++
> std::unique_lock locker(lock);
> cond.wait(locker, [&] { return done; });
> return initialized ? 0 : -EIO;
> ```

Ok

**Comment by fengchunsong** [comment]
@tchaikov please help to review again,thanks.

File: src/msg/async/dpdk/dpdk_rte.cc [hunk 3]
```diff
@@ -137,17 +137,21 @@ namespace dpdk {
       if (!rte_initialized) {
         /* initialise the EAL for all */
         int ret = rte_eal_init(cargs.size(), cargs.data());
-        if (ret < 0)
-          return;
+	if (ret < 0) {
+          std::unique_lock<std::mutex> locker(lock);
```
Comment (tchaikov) [inline src/msg/async/dpdk/dpdk_rte.cc:141]: no need to specify the template argument, as, C++17 is able to deduce the template argument type. see https://en.cppreference.com/w/cpp/language/template_argument_deduction .

File: src/msg/async/dpdk/dpdk_rte.cc [hunk 4]
```diff
@@ -137,17 +137,21 @@ namespace dpdk {
...
+          done = true;
+          cond.notify_all();
+          return ret;
+	}
         rte_initialized = true;
       }
 
-      std::unique_lock<std::mutex> l(lock);
+      std::unique_lock<std::mutex> locker(lock);
```
Comment (tchaikov) [inline src/msg/async/dpdk/dpdk_rte.cc:149]: ditto.

File: src/msg/async/dpdk/dpdk_rte.cc [hunk 5]
```diff
@@ -156,10 +160,9 @@ namespace dpdk {
         }
       }
     });
-    std::unique_lock<std::mutex> l(lock);
-    while (!done)
-      cond.wait(l);
-    return 0;
+    std::unique_lock<std::mutex> locker(lock);
```
Comment (tchaikov) [inline src/msg/async/dpdk/dpdk_rte.cc:163]: ditto.

**Review by tchaikov — APPROVED** [review]

**Pull request merged**